packages/common/src/downloads/_internal/file-url-fetchers/fetchHubApiDownloadFileUrl.ts

```diff
@@ -0,0 +1,222 @@
+import HubError from "../../../HubError";
+import { getProp } from "../../../objects/get-prop";
+import {
+  ArcgisHubDownloadError,
+  DownloadOperationStatus,
+  IFetchDownloadFileUrlOptions,
+  ServiceDownloadFormat,
+  downloadProgressCallback,
+} from "../../types";
+
+/**
+ * @private
+ * Fetches a download file url from the Hub Download API
+ *
+ * NOTE: The Hub Download API only works with a certain subset of Feature and Map services
+ * and performs different operations (i.e., calling createReplica or paging the service's
+ * features) depending on the service type and capabilities.
+ *
+ * This function does it's best to abstract those differences and provide a consistent
+ * interface for downloading data from any service supported by the Hub Download API.
+ *
+ * @param options options for refining / filtering the resulting download file
+ * @returns a url to download the file
+ */
+export async function fetchHubApiDownloadFileUrl(
+  options: IFetchDownloadFileUrlOptions
+): Promise<string> {
+  validateOptions(options);
+  const requestUrl = getDownloadApiRequestUrl(options);
+  const { pollInterval, progressCallback } = options;
+  return pollDownloadApi(requestUrl, pollInterval, progressCallback);
+}
+
+function validateOptions(options: IFetchDownloadFileUrlOptions) {
+  const { layers = [] } = options;
+
+  // The Hub Download API currently requires a target layer to be specified
+  if (layers.length === 0) {
+    throw new HubError(
+      "fetchHubApiDownloadFileUrl",
+      "No layers provided for download"
+    );
+  }
+
+  // The Hub Download API currently only supports downloading one
+  // layer at a time, though it could allow multiple in the future
+  if (layers.length > 1) {
+    throw new HubError(
+      "fetchHubApiDownloadFileUrl",
+      "Multiple layer downloads are not yet supported"
+    );
+  }
+}
+
+/**
+ * @private
+ * Generates a URL to the Hub Download API that can be polled until the download is ready
+ *
+ * @param options options for refining / filtering the resulting download file
+ * @returns a download api url that can be polled
+ */
+function getDownloadApiRequestUrl(options: IFetchDownloadFileUrlOptions) {
+  const { entity, format, context, layers, geometry, where } = options;
+
+  const searchParams = new URLSearchParams({
+    redirect: "false", // Needed to get the download URL instead of the file itself
+    layers: layers[0].toString(),
```

Comment:
The interface for `IFetchDownloadFileUrlOptions` shows `layers` as being optional, so it's possible `layers[0].toString()` may throw an error.

Reply:
Perfect, makes sense, thanks!